### PR TITLE
Change warning() to automatically inject the stack, and add warningWithoutStack() as opt-out

### PR DESCRIPTION
**Pull request by gaearon**
#### Summary

This flips `warning()` to always include a component stack **by default** if it exists. It also adds a counterpart `warningWithoutStack()` for cases where you **don't** want to have a stack.

I opted to have two separate methods because `warning` arguments already have a meaning (they get interpolated). I decided against something like `warning.withoutStack()` because this proved to be a bit annoying in the past (e.g. `emptyFunction.*`), and would be a tiny bit more difficult to adopt our transforms to. I don't feel strongly about this though.

**This PR shouldn't contain any functional changes.** Only warnings without stacks were converted to use the new function. Warnings that already have stacks just lost their last argument because it's provided automatically now.

#### Why

* I want to make it easy for newly added warnings to always have stacks.
* This removes a bunch of `ReactDebugCurrentFrame` direct calls, imports, and wiring.
* This paves the way towards doing more in `warning()`, e.g. determining whether we're in strict mode. Which is something we'll want to expose in every warning so that we can get rid of internal blacklist.

#### How to Review

* Calls to `warning()` that had stacks now omit the last formatting argument.
* Calls to `warning()` that didn't have stacks now use `warningWithoutStack()`.
* There are a few infra changes that make me more confident in this, e.g. I added validation of the warning argument number again the count of `%s` interpolations in tests.

One notable change is that instead of formatting before calling `console.error()`, we now pass through the message and the arguments, e.g. `console.error('%s is a bad component', 'Foo')`. This works in all major browsers (including IE), and Node.js. Theoretically this might be a breaking change for somebody who was relying on filtering `console.error` calls by exact messages. But we already change warning messages between releases anyway, and this change should make it *easier* to filter messages because the original formatting string is now exposed before the interpolation.

#### Work Log

- [x] Introduce `warningWithStack()` that automatically includes a stack from the currently injected implementation.
- [x] Start converting callsites to use `warningWithStack()` and remove threading of `getStack` through utilities shared between SSR and client.
- [x] Figure out places like `assertValidProps` and how they need the frame. Why don't test fails in prod there?
- [x] Consider what to do for isomorphic callsites. They currently wouldn't work with `warningWithStack()` because it would create a circular dependency between `React` and `warningWithStack`.
- [x] Maybe flip them. Make `warning()` have a stack, and `warningWithoutStack()` its explicit counterpart.
- [x] Fail the build if we use warning with stack in a package that doesn't have React dep
- [x] Change our transforms to handle the new warning method too.
- [x] Fix Flow

<details>
<summary>Follow-up work</summary>

- [ ] Consider snapshotting warnings?
- [ ] Sometimes we deduplicate based on the stack. Should `warningWithStack()` just do it by default? Other ideas?
- [ ] Convert all renderer callsites that include the stack to `warningWithStack()`.
- [ ] Remove direct usage of `getCurrentFiberStackAddendum`. Always go through the central isomorphic module.
- [ ] Consider doing the same for invariants.
- [ ] Enforce that number of formatting arguments is expected in tests.
- [ ] Consider attaching stack to invariants too? Make debug frame non-DEV-only?
- [ ] Include strictness into the warnings. Integrate StrictMode warnings with this new infra.
- [ ] Ensure internal callsites that depend on `%s` interpolation having already happened don't break

</details>

### CODE REVIEW

**Comment by gaearon** [comment]
@sebmarkbage Not ready for review but feel free to flag if I'm going in some completely different direction than you imagined

**Comment by react-linter** [comment]
ReactDOM: size: 0.0%, gzip: -0.0%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: f762b3abb1d951b366f3c8a2354aeedb7fbc32ff...62795d4029534828da978f8558a40c46978d5662</p>


  
## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | 0.0% | +0.1% | 639.67 KB | 639.82 KB | 149.78 KB | 149.92 KB | UMD_DEV
**react-dom.production.min.js** | **0.0%** | **-0.0%** | **95.5 KB** | **95.5 KB** | **30.9 KB** | **30.89 KB** | **UMD_PROD**
react-dom.development.js | 0.0% | +0.1% | 635.8 KB | 635.94 KB | 148.59 KB | 148.73 KB | NODE_DEV
**react-dom.production.min.js** | **0.0%** | **-0.0%** | **95.48 KB** | **95.49 KB** | **30.41 KB** | **30.41 KB** | **NODE_PROD**
react-dom-test-utils.development.js | +0.2% | +0.2% | 42.86 KB | 42.94 KB | 11.92 KB | 11.94 KB | UMD_DEV
react-dom-test-utils.development.js | +0.2% | +0.2% | 42.58 KB | 42.66 KB | 11.85 KB | 11.88 KB | NODE_DEV
react-dom-unstable-native-dependencies.development.js | +0.1% | +0.1% | 59.99 KB | 60.07 KB | 15.88 KB | 15.9 KB | UMD_DEV
react-dom-unstable-native-dependencies.development.js | +0.1% | +0.1% | 59.66 KB | 59.75 KB | 15.75 KB | 15.77 KB | NODE_DEV
react-dom-server.browser.development.js | +0.1% | +0.5% | 99.27 KB | 99.34 KB | 26.23 KB | 26.36 KB | UMD_DEV
**react-dom-server.browser.production.min.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.5%** | **14.81 KB** | **14.84 KB** | **5.69 KB** | **5.72 KB** | **UMD_PROD**
react-dom-server.browser.development.js | +0.1% | +0.5% | 95.39 KB | 95.46 KB | 25.29 KB | 25.42 KB | NODE_DEV
**react-dom-server.browser.production.min.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.4%** | **14.71 KB** | **14.74 KB** | **5.62 KB** | **5.64 KB** | **NODE_PROD**
react-dom-server.node.development.js | +0.1% | +0.5% | 97.32 KB | 97.38 KB | 25.83 KB | 25.95 KB | NODE_DEV
**react-dom-server.node.production.min.js** | **:small_red_triangle:+0.2%** | **:small_red_triangle:+0.5%** | **15.51 KB** | **15.54 KB** | **5.92 KB** | **5.95 KB** | **NODE_PROD**
ReactDOM-dev.js | -0.0% | +0.1% | 642 KB | 642 KB | 146.9 KB | 147.01 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **-0.0%** | **-0.0%** | **274.91 KB** | **274.85 KB** | **51.56 KB** | **51.55 KB** | **FB_WWW_PROD**
ReactDOMServer-dev.js | -0.1% | +0.6% | 96.45 KB | 96.4 KB | 24.83 KB | 24.98 KB | FB_WWW_DEV
**ReactDOMServer-prod.js** | **:small_red_triangle:+1.4%** | **:small_red_triangle:+1.0%** | **31.97 KB** | **32.41 KB** | **7.78 KB** | **7.86 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | 0.0% | -0.0% | 96.57 KB | 96.58 KB | 30.81 KB | 30.8 KB | NODE_PROFILING
ReactDOM-profiling.js | -0.0% | -0.0% | 277.94 KB | 277.88 KB | 52.22 KB | 52.22 KB | FB_WWW_PROFILING

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by Arissablue** [comment]
thanks. 💘

File: packages/react-reconciler/src/__tests__/ReactNewContext-test.internal.js [hunk 1]
```diff
@@ -711,18 +709,15 @@ describe('ReactNewContext', () => {
...
-        'calculateChangedBits: Expected the return value to be a 31-bit ' +
-          'integer. Instead received: 4294967295',
-        {withoutStack: true}, // TODO: add a stack
-      );
-    }
+    expect(() => {
+      // Update
+      ReactNoop.render(<Context.Provider value={2} />);
+      ReactNoop.flush();
+    }).toWarnDev(
```
Comment (bvaughn) [inline packages/react-reconciler/src/__tests__/ReactNewContext-test.internal.js]: tiny nit: Might be nicer to move `ReactNoop.render` outside of the `expect` call.

File: scripts/jest/matchers/__tests__/toWarnDev-test.js [hunk 2]
```diff
@@ -147,5 +147,19 @@ describe('toWarnDev', () => {
...
+          console.error('Hi %s', 'Sara', 'extra');
+        }).toWarnDev('Hi', {withoutStack: true});
+      }).toThrow('Received 2 arguments for a message with 1 placeholders');
+
+      expect(() => {
+        expect(() => {
+          console.error('Hi %s');
+        }).toWarnDev('Hi', {withoutStack: true});
+      }).toThrow('Received 0 arguments for a message with 1 placeholders');
+    });
```
Comment (bvaughn) [inline scripts/jest/matchers/__tests__/toWarnDev-test.js]: Nice 👍

File: scripts/jest/matchers/toWarnDev.js [hunk 3]
```diff
@@ -23,16 +24,28 @@ const createMatcherFor = consoleMethod =>
...
 
+        let argIndex = 0;
+        format.replace(/%s/g, () => argIndex++);
+        if (argIndex !== args.length) {
+          lastWarningWithMismatchingFormat = {
+            format,
+            args,
+            expectedArgCount: argIndex,
+          };
+        }
```
Comment (bvaughn) [inline scripts/jest/matchers/toWarnDev.js]: Clever!

Could use an inline comment. Wasn't clear (to me at least) on first read what it was doing.

File: scripts/rollup/forks.js [hunk 4]
```diff
@@ -40,10 +40,23 @@ const forks = Object.freeze({
...
+      return new Error(
+        'Cannot use a module that depends on ReactSharedInternals ' +
+          'from "' +
+          entry +
+          '" because it does not declare "react" in the package ' +
+          'dependencies or peerDependencies. For example, this can happen if you use ' +
+          'warning() instead of warningWithoutStack() in a package that does not ' +
+          'depend on React.'
+      );
+    }
```
Comment (bvaughn) [inline scripts/rollup/forks.js]: Nice!

File: scripts/rollup/plugins/use-forks-plugin.js [hunk 5]
```diff
@@ -33,7 +33,9 @@ function useForks(forks) {
     const targetModule = forks[srcModule];
     resolvedForks.set(
       require.resolve(srcModule),
-      require.resolve(targetModule)
+      typeof targetModule === 'string'
+        ? require.resolve(targetModule)
+        : targetModule
```
Comment (bvaughn) [inline scripts/rollup/plugins/use-forks-plugin.js]: Hm? 🤔

**Pull request merged**